File: AGRA_2/hw04/bridges/bridges/tarjanBridges.py
"""
Algoritmo de Tarjan (Bridges)

"""

from sys import stdin

G = [[] for _ in range(50000)]
visited = [False for _ in range(50000)]
low = [-1 for _ in range(50000)]
p = [-1 for _ in range(50000)]
bridgesSet = set()
t, n = int(), int()

def bridgesTarjan():
    global low, visited, p
    for i in range(n):
        low[i] = visited[i] = p[i] = -1

    for i in range(n):
        if visited[i] == -1:
            bridgesAux(i)

def bridgesAux(v):
    global low, visited, p, bridgesSet, t
    t += 1
    visited[v] = low[v] = t

    for w in G[v]:
        if visited[w] == -1:
            p[w] = v
            bridgesAux(w);
            low[v] = min(low[v], low[w])

            #verificar si es un puente
            if low[w] > visitado[v]:
                bridgesSet.add((v, w))
                
        elif w != p[v]:
            low[v] = min(low[v], visitado[w])

def main():
    global G
    
    n, m = map(int, stdin.readline().split())

    for i in range(m):
        u, v = map(int, stdin.readline().split())
        G[u].append(v)
        G[v].append(u)

    bridgesTarjan()

    if len(bridgesSet) == 0:
        print("El grafo no tiene puentes")
    else:
        print("Total puentes:", len(bridgesSet))
        print(*bridgesSet)

main()

File: AGRA_2/hw04/bridges/bridges/tarjanBridges.cpp
/*
Algoritmo de Tarjan (Bridges)
Autor: Carlos Ramirez
Fecha: Abril 4 de 2020

 */

#include <vector>
#include <stack>
#include <iostream>
#include <set>

using namespace std;

int n, numSCC, t;
vector<vector<int> > adj(50000);
int visitado[50000];
int low[50000];
int padre[50000];
set<tuple<int, int> > bridges;

void bridgesAux(int);

void bridgesTarjan(){
  int i;

  for(i = 0; i < n; i++)
    low[i] = visitado[i] = padre[i] =-1;

  for(i = 0; i < n; i++)
    if(visitado[i] == -1)
      bridgesAux(i);
}

void bridgesAux(int v){
  int w;
  visitado[v] = low[v] = ++t;

  for(int i = 0; i < adj[v].size(); i++){
    w = adj[v][i];
    if(visitado[w] == -1){
      padre[w] = v;
      bridgesAux(w);
      low[v] = min(low[v], low[w]);

      //verificar si es un puente
      if(low[w] > visitado[v])
	bridges.insert(make_pair(v, w));
    }
    else if(w != padre[v])
      low[v] = min(low[v], visitado[w]);
  }
}

int main(){
  int m, i, aux1, aux2;

  cin >> n >> m;

  for(i = 0; i < m; i++){
    cin >> aux1 >> aux2;
    adj[aux1].push_back(aux2);
    adj[aux2].push_back(aux1);
  }

  bridgesTarjan();

  if(bridges.size() == 0)
    cout << "El grafo no tiene puentes" << endl;
  else{
    cout << "Total de Puentes: " << bridges.size() << endl;
    for(set<tuple<int, int> >::iterator it = bridges.begin(); it !=  bridges.end(); ++it)
      cout << " (" << it->first << ", " << it->second << ")";
    cout << endl;
  }
  
  return 0;
}
